### ISSUE TITLE
sql: cannot add a new sequence-populated column (or pg-compatible SERIAL) to an existing table

### ISSUE THREAD

**Issue by knz** [cockroachdb/cockroach#42508]
Today the column backfiller is unable to _execute_ sequence operations such as evaluating the `nextval()` function.

This makes it impossible to add a column that uses a sequence-based DEFAULT to an existing table.

For example:

```sql
> create table t(x int); insert into t(x) values (1), (2), (3);
> create sequence s;
> alter table t add column y int default nextval('s');
pq: nextval(): cannot backfill such sequence operation
```

Or:
```sql
> set experimental_serial_normalization=sql_sequence; -- for pg compat
>  alter table t add column y serial;
pq: nextval(): cannot backfill such sequence operation
```

**A large problem with this limitation is that the error is only detected at runtime, not at compile-time.** This means that an attempt to add such a column inside an explicit txn will violate atomicity with the dreaded XXA00 error (see #42061) 

```sql
> begin; 
  insert into t(x) values (4); 
  alter table t add column y int default nextval('d'); 
  commit;
pq: transaction committed but schema change aborted with error: (42P15): nextval(): cannot backfill such sequence operation
HINT: Some of the non-DDL statements may have committed successfully, but some of the DDL statement(s) failed.
Manual inspection may be required to determine the actual state of the database.
--
See: https://github.com/cockroachdb/cockroach/issues/42061
> select * from t;
  x
+---+
  1
  2
  3
  4
(4 rows)
```

(the INSERT was committed, but the ALTER failed: atomicity anomaly)

**Comment by knz** [comment]
cc @jordanlewis @RaduBerinde @lucy-zhang for triage.

**Comment by knz** [comment]
Also I'm thinking about this a bit and I really don't see why the backfiller cannot evaluate sequence operations. They are non-transactional, so it should really be possible. @jordanlewis what do you think?

**Comment by lucy-zhang** [comment]
Would it be possible to detect sequence-populated columns at compile time, and backport a fix using that approach? As I understand it, #42509 continues to detect the problem only at runtime (though of course it would still be an improvement).

**Comment by knz** [comment]
I agree with your proposal.

That said, the main change in #42509 is removing the error flattening, to ensure that the more specific error codes bubble up no matter what. I think this should be merged first before the improvement you sugggest.

**Comment by jseldess** [comment]
This known limitation was documented some time ago: https://www.cockroachlabs.com/docs/stable/known-limitations.html#adding-a-column-with-certain-default-values. I'll add a link to this issue. Need to confirm if the language is still correct and sufficient.

**Comment by Bessonov** [comment]
Faced today in the same issue. [Accidentally discovered, that it can be done.](https://github.com/cockroachdb/cockroach/issues/36646#issuecomment-481264342). Furthermore, with

```sql
update mytable set myseq = nextval('seq') where myseq is null;
```
existing fields get filled. It would be nice to have that in docs.

**Comment by akss1** [comment]
For example, the described bug leads to [this](https://github.com/cockroachdb/cockroach/issues/36646#issuecomment-916064653).

**Comment by postamar** [comment]
Implementing ADD COLUMN in the declarative schema changer will help here, as the column backfiller will no longer be used.

**Comment by ajwerner** [comment]
The index backfiller is not any better equipped to interact with sequences.

**Comment by a-eid** [comment]
any updates on this ?

**Comment by z0mb1ek** [comment]
wait too
